### DISCUSSION TITLE
Discrepancy in YOLOv11 Layer Count and Performance Between Cloned Repo and Pip-Installed Package

### DISCUSSION THREAD

**Discussion by AnirudhSanthana** [ultralytics/ultralytics#20034] [Q&A]
Hello Ultralytics Team, @glenn-jocher 

I'm encountering a significant discrepancy in the layer count and resulting performance when training YOLOv11 using two different methods. Here are the details:

Method 1: Cloned Repository (Editable Installation)

Setup while cloning the repo

_from google.colab import drive
drive.mount('/content/drive')

%cd /content
**!git clone https://github.com/ultralytics/ultralytics.git**
%cd ultralytics

!pip install -e .
!pip install --upgrade torch_

Ran training using:

_from ultralytics import YOLO
model = YOLO('/content/ultralytics/ultralytics/cfg/models/11/yolo11.yaml')
results = model.train(
    data='/content/drive/MyDrive/dataset/data.yaml',
    epochs=100,
    batch=16,
    imgsz=320
)_

Summary:
YOLO11n summary: 181 layers, 2,592,180 parameters, 2,592,164 gradients, 6.5 GFLOPs
Validation results: mAP50 ~0.726, mAP50-95 ~0.622

Method 2: Pip-Installed Version

Setup:

Installed via pip:

_!pip install ultralytics_

Ran training using 

_!yolo detect train data=/content/drive/MyDrive/dataset/data.yaml model=yolo11n.pt epochs=100 imgsz=320 batch=16_

Summary:
YOLO11n summary: 319 layers, 2,592,180 parameters, 2,592,164 gradients, 6.5 GFLOPs
Validation results: mAP50 ~0.888, mAP50-95 ~0.825

Questions:

1) Layer Count Difference:
Why does the cloned repository version show 181 layers while the pip-installed version shows 319 layers, despite having the same parameter count and GFLOPs?

2) Architectural Modifications:

If I want to perform architectural modifications (e.g., integrating attention modules like CBAM at specific locations), what is the recommended approach? Is it better to work with the editable cloned version or is there a way to override these configurations while using the pip-installed package and if so how do we do it?

I appreciate your guidance and any insights you can provide on these discrepancies.

Thank you for your time 
Best regards,
Anirudh
<img width="491" alt="image" src="https://github.com/user-attachments/assets/55a86610-af54-47a8-a556-05ce74724707" />
<img width="486" alt="image" src="https://github.com/user-attachments/assets/40290855-baea-4596-9a38-1b165ac98557" />

**Additional Note on Performance:**
When I ran the pip-installed version using _pip install -U ultralytics,_ I obtained the following validation performance:

mAP50: 0.893, mAP50-95: 0.831

This performance improvement compared to the cloned version is extremely confusing given that both approaches show the same total parameter count and GFLOPs but differ significantly in layer count and consequently in detection metrics.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @AnirudhSanthana, thank you for reaching out and for your detailed report! 🚀

We recommend checking out the [Ultralytics Docs](https://docs.ultralytics.com/) for guidance on training, customization, and architecture modifications. Specifically, you may find the [Python Usage](https://docs.ultralytics.com/usage/python/) and [CLI Usage](https://docs.ultralytics.com/usage/cli/) sections helpful for understanding the differences between training methods. For architecture adjustments, the [Model Configuration](https://docs.ultralytics.com/usage/cfg/) guide is a great starting point.

### 🐛 Bug Report
If this is a bug, we kindly request a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate the discrepancy in layer counts and performance. Please ensure that all relevant files, configurations, and steps are included.

### Customization ❓
For architectural modifications, such as integrating CBAM or other modules, we recommend using the editable cloned repository (`pip install -e .`) for maximum flexibility. However, if you prefer working with the pip-installed version, it is possible to override configurations by specifying custom `.yaml` files or using the Python API. Feel free to share more details about your intended changes if you need further assistance.

### Community and Support
Join the Ultralytics community to discuss your findings and share insights:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time conversations
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to connect with the community

### Upgrade
Ensure you are using the latest version of `ultralytics`. Upgrade with the following command to ensure all recent fixes and updates are applied:

```bash
pip install -U ultralytics
```

### Verified Environments
YOLO can be run in various verified environments. For ease of use, consider running your experiments in the following environments that come pre-configured with all dependencies:
- **Notebooks** with free GPUs: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, ensuring all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) work as expected across systems.

This is an automated response to ensure you get the fastest support possible. An Ultralytics engineer will review your query and provide further assistance soon. Thank you for your patience! 😊

**Reply by glenn-jocher** [reply]
Hi @AnirudhSanthana, the significant performance difference typically arises because loading `yolo11n.pt` utilizes pre-trained weights, whereas loading the `.yaml` file initiates training from scratch.

The layer count discrepancy in the summary output is likely a reporting difference between loading methods (`.pt` vs `.yaml`) and doesn't reflect a change in the actual model complexity, given the identical parameters and GFLOPs.

For making architectural changes like adding attention modules, modifying the relevant `.yaml` configuration files within your cloned repository (installed via `pip install -e .`) is indeed the standard and most flexible method.

**Comment by AnirudhSanthana** [comment]
Hi @glenn-jocher

Thanks for your response, I understand that the performance difference mainly comes from the fact that loading _yolo11n.pt_ uses pre-trained weights, whereas loading the YAML starts training from scratch. I have one more question regarding architectural modifications with custom modules (e.g., adding CBAM):

When I customize YOLO using my custom YAML (which integrates attention modules), for a fair comparison with the standard model, should I:

- Train both the standard YOLO and my modified YOLO from scratch (i.e., without using any pre-trained weights), or
- Fine-tune my customized model by first loading pre-trained weights (from, say, yolo11n.pt), saving those weights, and then retraining (or fine-tuning) using those weights—so that both models benefit from pre-training?

In other words what's the best practice for benchmarking my architectural modifications against the standard YOLO model? I want to ensure that the comparison is fair and that any performance gains (or losses) can be attributed solely to the modifications like CBAM rather than differences in weight initialization.

Id appreciate your suggestions on the recommended approach and any best practices you might have.
Thank you for your guidance!

**Reply by glenn-jocher** [reply]
Hi @AnirudhSanthana,

For the fairest comparison focused purely on the architectural modifications (like adding CBAM), the standard practice is to train both the baseline model (standard YOLO11) and your modified model from scratch on your target dataset. This ensures that performance differences are attributable to the architectural changes rather than the influence of pre-trained weights.

**Comment by Y-T-G** [comment]
The first one is correct. And the pip installation should also show the same count if you're using the latest version.

https://github.com/ultralytics/ultralytics/pull/19202
